### ISSUE TITLE
PHP Fatal error: Non-static method cannot be called statically

### ISSUE THREAD

**Issue by dylanglass6300** [philbertphotos/osticket-multildap-auth#51]
osTicket - v1.16.2
PHP - 8.0.18
Apache - 2.4.53

I've added the plugin, and it shows in my list of currently installed plugins. When I click on the name to configure it, the options area is empty, pictured below:

![image](https://user-images.githubusercontent.com/105012534/166987810-eab917ed-90ab-42c3-8a6c-9667b9fb40ea.png)

In my apache error log, the error I am getting is:

 [Thu May 05 14:15:06.025931 2022] [php:error] [pid 772:tid 1872] [client ::1:55923] PHP Fatal error:  Uncaught Error: Non-static method LdapMultiAuthPlugin::DateFromTimezone() cannot be called statically in C:\\xampp\\htdocs\\osticket\\include\\plugins\\multi-ldap\\config.php:28\nStack trace:\n#0 C:\\xampp\\htdocs\\osticket\\include\\plugins\\multi-ldap\\config.php(278): LdapMultiAuthPluginConfig->getschedule()\n#1 C:\\xampp\\htdocs\\osticket\\include\\class.plugin.php(12): LdapMultiAuthPluginConfig->getOptions()\n#2 C:\\xampp\\htdocs\\osticket\\include\\class.plugin.php(485): PluginConfig->__construct('9')\n#3 C:\\xampp\\htdocs\\osticket\\include\\staff\\plugin.inc.php(5): Plugin->getConfig()\n#4 C:\\xampp\\htdocs\\osticket\\scp\\plugins.php(66): require('C:\\\\xampp\\\\htdocs...')\n#5 {main}\n  thrown in C:\\xampp\\htdocs\\osticket\\include\\plugins\\multi-ldap\\config.php on line 28, referer: http://localhost/osticket/scp/plugins.php

**Comment by philbertphotos** [comment]
I have not tested in PHP 8.0 I would have to make changes for it to work in that version.

**Comment by philbertphotos** [comment]
For now if you comment out line 278 and 279 should correct the issue for now till I create a fix.
			'sync_schedule_show' => new SectionBreakField(array(
				//'label' => $__('Next schedule: ' . $this->getschedule()) ,
				//'hint' => $__('Last run: ' . $this->getlastschedule()) ,
			)) ,

**Comment by dylanglass6300** [comment]
This worked to allow me to configure the plugin but I ran into more issues afterwards. I went ahead and downgraded and that allowed me to get through configuration and save without errors. However, now I get a different issue. It has not yet synced for the first time and says 

"Next schedule: December 31 1969 7:00 pm
Last run: December 31 1969 7:00 pm"

**Comment by philbertphotos** [comment]
@dylanglass6300 let me look at somethings and do it from scratch to see what is wrong. You will not need to comment out the functions if you downgraded PHP.

**Comment by philbertphotos** [comment]
@dylanglass6300  have you tried this? I remember adding it to correct this issue.
![image](https://user-images.githubusercontent.com/2892474/167983516-387fc71a-8be2-44c9-b2a0-52bdbbf626b5.png)

**Comment by catalinchertes** [comment]
I'm running into the same issues with php 8.0, will you make it compatible or is it a big headache? Thank you.

**Comment by philbertphotos** [comment]
@catalinchertes and @dylanglass6300  actively working on this should be done by Friday.

**Comment by SamehElsakka** [comment]
I have the same issue it shows Next schedule: January 01 1970 3:00 am and the reset/update schedule is ticked ,also the sync users button not working, i am using php7.4 and osticket version 1.15.2

**Comment by philbertphotos** [comment]
Yea I am working on this and should have it fixed by today ... hopefully.

I think this bug is because when I make changes I almost never install the plugin from scratch and it not creating the proper SQL tables to work properly

**Comment by philbertphotos** [comment]
Found the problem it was a missing database key ... this is a bug fix. I will update it a few more times after doing a few more tests but this works after removing it and reinstalling the plugin several times.

**Comment by PaysenPetersen** [comment]
I uninstalled the old version and installed the new version. Now you can save the data again and there are no more error messages.
But the Sync button has no function for me and the LDAP users are also not syncralized. Does an additional library have to be installed? Or are special rights for a directory or database table?

I am currently users Ubuntu 20.04 and PHP 8. The Osticket LDAP authentication works.

I have a total of 3 ad domains Windows 2012 R2. Can you please see why the Sync doesn't work? In some screening, I always see an assignment table that is not shown for me.

**Comment by philbertphotos** [comment]
That part works but only if you use cron jobs … thats the only time the sync activates … i should put that in the readme. 
3 AD domains is nice .. i tested up to 4 but actively use two.

**Comment by philbertphotos** [comment]
You may manually have to copy that missing file to the scp folder some how it failed to copy over…

**Comment by PaysenPetersen** [comment]
Hello, the syncronization doesn't work for me either. The Sync button currently has no function.

When clicking on Sync Users, he gets stuck here.

![image](https://user-images.githubusercontent.com/67009826/169512923-ef416b7f-53df-47a9-a4f7-9402490590ad.png)

**Comment by philbertphotos** [comment]
@PaysenPetersen I working on all the bugs right now. Mostly working on PHP 8.0 compatibility had to do a partial rewrite of the code what a pain but its almost finished and doing a few enhancements in tow. 
The update will be released sometime today in a separate folder until I can make sure its backward compatible and I will make sure the syncing works again.

**Comment by philbertphotos** [comment]
As for the sync button I am trying to remember if I removed the code because of some previous issue. I will check again before the release.

**Comment by catalinchertes** [comment]
Thank you so much for dedicating your time and working on this!!

**Comment by philbertphotos** [comment]
@catalinchertes yea I need to fix it and a few other things that have been nagging me ... currently cleaning up code and removed BS and updating it to work with PHP 8.0 and backwards. Almost done. Hoping to release it soon.

**Comment by philbertphotos** [comment]
Ok guys try the new update and let me know..

Tested it on both PHP 7.4 and 8.1 and both current versions of Osticket 

Sync is working for me (it always did) but now I think I work out the kinks to make sure it adds all the SQL tables it needs.

**Comment by SamehElsakka** [comment]
Thanks for your time and effort but  i installed the new update and still the same issues ,the sync schedule showing January 01 1970 3:00 am even when ticking on Reset/Update schedule on save and the Sync button still have no function at all.

**Comment by PaysenPetersen** [comment]
Hello, I also installed the update. I also had the problem of Samehelsokka. Despite reinstallation, I had to replace the file /var/www/osticket/upload/scp/sync_mldap.php from the plugin directory. Then it works too. Otherwise I got an HTTP 500 error.

Thank you for your time and effort.

However, I still have a question about the spelling of the password. How do you have to enter the password correctly if you have several domains and thus several passwords.

**Comment by PaysenPetersen** [comment]
I'm just testing.

The agents can authenticate themselves with the 1 domain via LDAP.
However, customers are not syncralized.


![image](https://user-images.githubusercontent.com/67009826/169789202-8e3e9eb5-5219-40e0-8e9f-37aacf54ff8b.png)

![image](https://user-images.githubusercontent.com/67009826/169789572-291b02da-0be8-4e19-b1fe-6c3597c0f37d.png)

**Comment by philbertphotos** [comment]
@PaysenPetersen THIS IS PROGRESS! Ok let me check if I can find out why... please check your error and cron logs and send me any errors you may have gotten.

I will check into it and get a update out today ... should be an easy bug to correct.

**Comment by PaysenPetersen** [comment]
I just exchanged the version.

![image](https://user-images.githubusercontent.com/67009826/169852500-2cd1bfae-dab0-4d44-aec2-373fee324712.png)

No remote user appears for me when it comes to the new system. The cron job is running

**cron.log**
May 23 14:57:24 ticketsytem cron[39089]: (CRON) INFO (pidfile fd = 3)
May 23 14:57:24 ticketsytem cron[39089]: (CRON) INFO (Skipping @reboot jobs -- not system startup)
May 23 15:09:01 ticketsytem CRON[39164]: (root) CMD (  [ -x /usr/lib/php/sessionclean ] && if [ ! -d /run/systemd/system ]; then /usr/lib/php/sessionclean; fi)
May 23 15:17:01 ticketsytem CRON[40511]: (root) CMD (   cd / && run-parts --report /etc/cron.hourly)

I have currently only configured one domain.
The syncronization and searches do not work yet.

**Comment by PaysenPetersen** [comment]
Is it correct that both LDAP modules are active? The standard module is not filled out.
![image](https://user-images.githubusercontent.com/67009826/169854753-8fb14f02-d46d-4097-838a-3a4ff6c22638.png)


![image](https://user-images.githubusercontent.com/67009826/169853785-867c08c0-a09a-4b3f-b349-400ab31a8b08.png)

**Comment by philbertphotos** [comment]
@PaysenPetersen  this is concerning... its not searching your LDAP/AD so let me create some visual checks so we can see what is going on. The fact that it authenticates your agents is interesting are you sure you pointing those agents LDAP?
![image](https://user-images.githubusercontent.com/2892474/169884619-066da22d-720a-4248-998e-48f978aa1a4c.png)
The issue here is AD is not connecting correctly it seems.

**Comment by philbertphotos** [comment]
Added some debug code and connection test button. Let me know what happens.

![image](https://user-images.githubusercontent.com/2892474/169912711-309fcfdb-e142-4d91-b50e-8df3dc19536b.png)

**Comment by PaysenPetersen** [comment]
1. for me the Check Connection Button has no function.
2. For a colleague, I have set LDAP especially. This works.
![image](https://user-images.githubusercontent.com/67009826/169984542-792f02a2-b32a-4b83-833f-05d1f2d1b2d2.png)
![image](https://user-images.githubusercontent.com/67009826/169984616-bf4dec80-9af1-4938-8ae6-1cb04d1b6ee1.png)
3. This Error Found in osticket_Error.log
[Tue May 24 10:31:50.068833 2022] [php:warn] [pid 60057] [client 172.18.97.87:53242] PHP Warning:  Undefined array key "sync" in /var/www/osTicket/upload/scp/sync_mldap.php on line 2, referer: http://ticketsystem.jlw-holding.local/scp/plugins.php?id=17

**Comment by philbertphotos** [comment]
The LDAP service is it Windows AD?

**Comment by PaysenPetersen** [comment]
Could it be that the comma is a problem in the CN? We use the German version.

**Comment by philbertphotos** [comment]
@PaysenPetersen could be can you send me a facsimile of the CN that way I can see how different it could be.

**Comment by catalinchertes** [comment]
Hey @philbertphotos, I downloaded the updated version. To have the check conectivity function working and save button, I had to copy the file sync_mldap.php to /scp folder manually. Now it says that it connected correctly to my domain but the sync function returns 0 users. What could I be doing wrong? Thanks!

**Comment by catalinchertes** [comment]
Same issue here, only agents can authentificate by LDAP, users can't and won't sync.

**Comment by catalinchertes** [comment]
For me the problem was in the Search User field, I solved that problem by writing it like this: cn=myaduser,cn=Users,dc=Mydomain,dc=local. It now shows that 150 users synced. But I have another problem, If I search for a user that is not registered but exists in my Active Directory, I can see it as (remote) user, but when I click on the user I get this error: Failed to load resource: the server responded with a status of 500 (Internal Server Error) on /scp/ajax.php/select/auth(and a long string with the CN where the user is located and my domain). And after getting this error I get logged out of my account.

**Comment by philbertphotos** [comment]
@catalinchertes  never thought the issue could be the search string .. sounds like the same issue @PaysenPetersen let me look at a few things its hard to pin down since I cant duplicate the problem...never had it and I always sync fine.

**Comment by philbertphotos** [comment]
Ok fixed the issue in line 877 and it looks like you now syncing users. 
As for the long string error I think its because of the bug I introduced and hopefully fixed.

**Comment by philbertphotos** [comment]
As for your search string I dont know how what you do works its not a LDAP string like this 
(&(objectCategory=person)(objectClass=user)(|(sAMAccountName={q}*)))

**Comment by e-p-s** [comment]
for me same problem. when i put it into pluginsfolder osticket didnt start anymore. no gui. apache error

**Comment by philbertphotos** [comment]
@e-p-s show me the error... in the PHP log.

**Comment by catalinchertes** [comment]
After enabling the plugin, now my user search function is broken and I get this in the console, even after disabling the plugin now the search functionaly is broken on my osticket.
![image](https://user-images.githubusercontent.com/57810465/170299216-48dc1d55-5827-4727-b1ec-b5fd76db6b43.png)

The other problem I face is that only Agents can authentificate by LDAP, users get the error message: "Contact the administrator to have an account created for you" This users were already created in Osticket and exist in AD.

**Comment by philbertphotos** [comment]
You need to turn this on like this or user login and creation. existing users are automatically synced/created even if they exist .... but let me test that again I have not in years since I never had a need to do so.
![image](https://user-images.githubusercontent.com/2892474/170315100-cf29004a-c794-485d-b534-e678bc80f018.png)

**Comment by philbertphotos** [comment]
looks like something changed in this version for client logins so let me check a few things should be an easy fix.

**Comment by e-p-s** [comment]
/var/log/apache2/error.log
`
PHP Warning:  Module "imap" is already loaded in Unknown on line 0
PHP Warning:  Module "ldap" is already loaded in Unknown on line 0
PHP Warning:  Module "mbstring" is already loaded in Unknown on line 0
[Tue May 31 13:56:44.161818 2022] [mpm_prefork:notice] [pid 583] AH00163: Apache/2.4.53 (Debian) configured -- resuming normal operations
[Tue May 31 13:56:44.161941 2022] [core:notice] [pid 583] AH00094: Command line: '/usr/sbin/apache2'
[Tue May 31 13:58:05.504699 2022] [php:error] [pid 655] [client 172.21.61.190:53471] PHP Fatal error:  Cannot make static method LDAPMultiAuthentication::getconfig() non static in class SyncLDAPMultiClass in /var/www/html/scp/sync_mldap>
[Tue May 31 13:59:37.371670 2022] [php:error] [pid 654] [client 172.21.61.190:53572] PHP Fatal error:  Cannot make static method LDAPMultiAuthentication::getconfig() non static in class SyncLDAPMultiClass in /var/www/html/scp/sync_mldap>
[Tue May 31 14:27:24.676075 2022] [php:error] [pid 657] [client 172.21.61.190:55030] PHP Fatal error:  Cannot make static method LDAPMultiAuthentication::getconfig() non static in class SyncLDAPMultiClass in /var/www/html/scp/sync_mldap>

`

**Comment by philbertphotos** [comment]
@PaysenPetersen ohhhh that makes sense now. Yea yea I think I know the issue.
As for a CN with spaces this is my CN and it works
CN=VIDE Service Account,OU=Service Accounts,DC=local,DC=com.

If the search works then the syncing should work also since its the same (or should be but I think I know why it wont so let me comb through the code for the differences.
Glad you get some of the other issue ironed out.

As for the time stuff I have no idea how to make it compatible for all time formats.

**Comment by PaysenPetersen** [comment]
Hello, I posted the code for the time problem at the top. Please test the code with you. I think you have an English version at the start. I installed your last version. However, the CN search with spaces does not work for me. Did you already patch? If you tested it again, I would test it again with me. Because at the moment I can't use the add on. If a non -registered user registers via the LDAP authentication, the name and the email address will not be fetched. As a result, there is an error when completing the ticket position.

If you ask the language and like it on Germany, you could largely deal with the problem. Unfortunately I have been out of programming for a long time. The adjustment had made a colleague who is currently not there.

In the screenshot you can still see 2 \\ in the CN. Could that still be a problem?

In addition, I do not see any imported data record in the customer overview.

**Comment by philbertphotos** [comment]
@PaysenPetersen  give me a few days going over the issue.
Does your CN have German characters in it? You blacked it out.

**Comment by philbertphotos** [comment]
@PaysenPetersen  Ok found the bug and was able to repeat it when a new user tries to login.

**Comment by philbertphotos** [comment]
Ok fixed both problems and uploading the changes tonight.

It was my fault since introduced a bug that broke the automatic user creation when fixing/enhancing other parts of the code. Never bother to check a new user from scratch.

**Comment by philbertphotos** [comment]
@PaysenPetersen I notice the CD you show with a space its shows "\\" is that how the German characters handles spaces? Seems a bit weird/non standard.
I think I fixed all the bugs and will close this issue once you confirm it works for you and the issues still pending be moved as future enhancements.
